misc/multistream-select/src/listener_select.rs

```diff
@@ -72,7 +73,14 @@ where
     // TODO: It would be nice if eventually N = Protocol, which has a
     // few more implications on the API.
     protocols: SmallVec<[(N, Protocol); 8]>,
-    state: State<R, N>
+    state: State<R, N>,
+    /// Whether the listener rejected a proposed protocol.
+    ///
+    /// If the listener reads garbage or EOF after having rejected
+    /// a protocol, the dialer is likely using `V1Lazy` and
+    /// negotiation must be considered failed, but not with a
+    /// protocol violation or I/O error.
+    rejected_protocol: bool,
```

Comment:
Instead of tracking whether one has ever rejected a protocol, would it make more sense to track whether the previous message send was a `NotAvailable`?

Sample scenario:

1. The listener rejects a protocol by sending `NotAvailabel` thus setting `rejected_protocol` to `true`.

2. The dialer then sends a `ListProtocols`.

3. The listener returns a `Protocols` message.

4. The dialer sends garbage down the line.

5. The listener terminates with a `NegotiationErr::Failed` given that `rejected_protocol` is still set to `true` instead of reporting the protocol violation.

Reply:
Sounds like a sensible refinement. I will try to do that.

Reply:
How about https://github.com/libp2p/rust-libp2p/pull/1871/commits/de3f0478013d629cbc77189f710be83dd326aa3f?